### PR TITLE
Refactor organizations

### PR DESCRIPTION
**Pull request by sampaccoud**
## Purpose

This is counter proposal for refactoring organizations and try to make it simpler and more powerful

## Proposal

Organizations were refactored with the following in mind:
- It's easier to start from scratch,
- Try not to rely on PageExtensions but have our own separate Organization model. The trick was to introduce a OneToOne relation from the Organization to a normal Page so we can still make `my_page.organization.logo.url` while relying on pure DjangoCMS pages,
- Add a wizard specific to adding a new Organization page, which creates an Organization model and links it to the newly created page,

Note: ended-up refactoring also the `create_cms_data` command because its design was making it hard to understand, impossible to `really test` and very slow to `test a little bit`,

### CODE REVIEW

File: apps/organization/tests/test_models.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        OrganizationFactory(code='a'*100)
+        with self.assertRaises(DataError):
+            OrganizationFactory(code='a'*101)
+
+    def test_organization_fields_not_required(self):
+        """
+        The `page` and `logo` are not required on the model
+        """
+        page = OrganizationFactory(page=None, logo=None)
+        self.assertTrue(page.pk is not None)
```
Comment (jmaupetit) [inline apps/organization/tests/test_models.py]: `self.assertIsNotNone(page.pk)`?

File: apps/organization/tests/test_models.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        self.assertTrue(page.page is None)
```
Comment (jmaupetit) [inline apps/organization/tests/test_models.py]: `self.assertIsNone(page.page)`?

File: apps/organization/tests/test_models.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        self.assertFalse(bool(page.logo))
```
Comment (jmaupetit) [inline apps/organization/tests/test_models.py]: `self.assertIsNone(page.logo)`?

File: apps/organization/tests/test_models.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    def test_organization_str_no_page(self):
+        """
+        The str representation should not include mention of the page if it does not exist
+        """
+        organization = Organization(page=None, name='Sorbonne', code='SOR')
+        self.assertEqual(str(organization), 'Sorbonne (SOR)')
+
+    def test_organization_str_with_page(self):
+        """
+        The str representation should include mention of the page if it exist
```
Comment (jmaupetit) [inline apps/organization/tests/test_models.py]: `s/exist/exists/`

File: apps/organization/tests/test_wizards.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+from ..cms_wizards import OrganizationWizardForm
+from ..factories import OrganizationFactory
+
+
+class OrganizationCMSWizardTestCase(CMSTestCase):
+    """Organization tests"""
+
+    def test_organization_create_wizards_list(self):
+        """
+        The wizard to create a new Organization page should be present of the wizards list page
```
Comment (jmaupetit) [inline apps/organization/tests/test_wizards.py]: `s/of/on/`?

File: apps/organization/tests/test_wizards.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        url = reverse('cms_wizard_create')
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
+        content = response.content.decode()
+        self.assertIn('Create a new Organization instance.', content)
+        self.assertIn('New Organization page', content)
+
+    def test_organization_wizard(self):
+        """
+        A user we sufficient access rights should be able to create an organization via the
```
Comment (jmaupetit) [inline apps/organization/tests/test_wizards.py]: `s/we/with/`

File: apps/organization/tests/test_wizards.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+        self.assertEqual(Page.objects.drafts().count(), 2)
+        self.assertEqual(Page.objects.count(), 2)
+        self.assertEqual(page.get_title(), 'My title')
+        expected_values = {
+            'name': 'My title',
+            'code': 'my-title',
+            'page': page,
+            'logo': None
+        }
+        self.assertTrue(page.organization, expected_values)
```
Comment (jmaupetit) [inline apps/organization/tests/test_wizards.py]: `assertTrue` and not `assertEqual`?

File: apps/core/helpers.py [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Helpers that can be useful throughout the whole project
+"""
+from django.template.defaultfilters import slugify
+
+from cms.api import add_plugin, create_page, create_title
+
+
+def create_i18n_page(content, is_homepage=False, **kwargs):
+    """
+    Creating an multilingual page is not straightforward so we thought we should have a helper
```
Comment (jmaupetit) [inline apps/core/helpers.py]: `s/an/a/`

Comment (jmaupetit) [inline apps/core/helpers.py]: Can you document the expected `content` argument?

File: apps/core/helpers.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        # Publish page in each additional language
+        if kwargs.get('published') is True:
+            page.publish(language)
+
+    # Add a plugin for each language (including the first language this time...)
+    placeholder = page.placeholders.get(slot='maincontent')
+    for language in content.keys():
+        add_plugin(
+            placeholder=placeholder,
+            plugin_type='TextPlugin', language=language,
```
Comment (jmaupetit) [inline apps/core/helpers.py]: The `language` parameter should be set in a new line :wink:

File: apps/core/management/commands/create_cms_data.py [hunk 10]
```diff
@@ -1,143 +1,110 @@
...
     cms_models.CMSPlugin.objects.all().delete()
     cms_models.Placeholder.objects.all().delete()
-    OrganizationPage.objects.all().delete()
+    Organization.objects.all().delete()
 
 
 def create_cms_data():
-    """Create base CMS data"""
+    """
+    Automatically create a base site for working on the project
```
Comment (jmaupetit) [inline apps/core/management/commands/create_cms_data.py]: `s/for working/to work/`?

File: richie/settings.py [hunk 11]
```diff
@@ -182,9 +182,11 @@ class Base(DRFMixin, ElasticSearchMixin, Configuration):
     # Use "en" as default as it is the language that is most likely to be spoken by any visitor
     # when their preferred language, whatever it is, is unavailable
     LANGUAGES = (
-        ('en', _('en')),
-        ('fr', _('fr')),
+        ('en', _('English')),
+        ('fr', _('French')),
```
Comment (jmaupetit) [inline richie/settings.py]: :medal_sports:

File: apps/core/tests/test_pages.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+        )
+        # Get the root page in french...
+        root = Page.objects.get_home()
+        response = self.client.get(root.get_absolute_url('fr'))
+        self.assertEqual(200, response.status_code)
+        # ... and make sure the page menu is present in french on the page
+        self.assertIn(content['fr'], response.rendered_content)
+
+        # Get the root page in english...
+        root = Page.objects.get_home()
```
Comment (jmaupetit) [inline apps/core/tests/test_pages.py]: You can safely remove this line (`root` has already been initialized with the same value).

File: Makefile [hunk 13]
```diff
@@ -97,5 +97,5 @@ watch-ts: ## watch changes in TypeScript files
 .PHONY: watch-ts
 
 help:
-	@grep -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
+	@grep -E '^[a-zA-Z0-9_-]+:.*?## .*$$' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
Comment (jmaupetit) [inline Makefile]: :+1:

File: setup.cfg [hunk 14]
```diff
@@ -10,6 +10,9 @@ exclude =
 
 [tool:pytest]
 addopts = -v --cov-report term-missing
+python_files =
+    test_*.py
+    tests.py
```
Comment (jmaupetit) [inline setup.cfg]: :+1:

File: apps/organization/tests/test_admin.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+        response = self.client.get(url, follow=True)
+        self.assertEqual(response.status_code, 200)
+        content = response.content.decode()
+        # Check that the page includes all our fields
+        self.assertIn(organization.name, content)
+        self.assertIn(organization.code, content)
+        self.assertIn(
+            '<td class="field-view_in_cms"><a href="/en/my-title/" target="_blank">View</a></td>',
+            content,
+        )
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organization/tests/test_admin.py]: Because I didn't know it exists lol. Done.

File: apps/organization/tests/test_admin.py [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+        user = UserFactory(is_staff=True, is_superuser=True)
+        self.client.login(username=user.username, password='password')
+        # Get the admin change view
+        url = reverse('admin:organization_organization_add')
+        response = self.client.get(url, follow=True)
+        self.assertEqual(response.status_code, 200)
+        content = response.content.decode()
+        # Check that the page includes all our fields
+        for field in ['page', 'name', 'code', 'logo']:
+            self.assertIn(field, content)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organization/tests/test_admin.py]: hardcoding the fields is superior for a test because it ensures that we don't add or remove fields unintentionally?

[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/organization/tests/test_admin.py]: I was pretty sure that you will answer this. This is a decent strategy. But still, the `assertIn` of a generic word as page, name or logo is too weak don't you think? This is why I proposed to assert at least `id="id_foo"`. WDYT?

[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organization/tests/test_admin.py]: Yes I took into account that change!

File: apps/organization/tests/test_models.py (hunk 3 above)
Comment (sampaccoud) [reply apps/organization/tests/test_models.py]: This one no. An imageField is not None and must be checked as a bool if I understood well.

File: apps/organization/tests/test_wizards.py (hunk 6 above)
Comment (sampaccoud) [reply apps/organization/tests/test_wizards.py]: Sorry #dyslexiaagain!

File: apps/organization/tests/test_wizards.py (hunk 7 above)
Comment (sampaccoud) [reply apps/organization/tests/test_wizards.py]: Wow why is this test working btw

Comment (jmaupetit) [reply apps/organization/tests/test_wizards.py]: :man_shrugging: duno

File: apps/core/helpers.py (hunk 9 above)
Comment (sampaccoud) [reply apps/core/helpers.py]: ... and sorted!

**Comment by sampaccoud** [comment]
@jmaupetit I added some line breaks in tests and fixed all the issues you raised. Feel free to look at it again and let me know if something is still not exactly the way you would like to see it?

File: apps/organizations/models.py [hunk 17]
```diff
@@ -0,0 +1,91 @@
...
+
+class Organization(models.Model):
+    """
+    The Organization model is used to represent and record entities that manage courses.
+    It could be a university or a training company for example.
+
+    This model should be used to record structured data about the organization whereas the
+    associated page object is where we record the less structured information to display on the
+    page to present the organization.
+    """
```
Comment (jmaupetit) [inline apps/organizations/models.py]: Why not giving this model a [`UUIDField`](https://docs.djangoproject.com/en/2.0/ref/models/fields/#uuidfield) as primary key? (I'm a big fan of uuids).

File: apps/organizations/models.py [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+    def save(self, *args, **kwargs):
+        """
+        Try to normalize the code a little for better uniqueness:
+            - all upper case,
+            - strip leading and trailing spaces,
+            - replace other spaces by a hyphen.
+        TODO: improve uniquess (find related Django blog post)
+        """
+        if self.code:
+            self.code = self.code.upper().strip().replace(' ', '-')
```
Comment (jmaupetit) [inline apps/organizations/models.py]: Why not using `slugify(self.code).upper() + uuid.uuid4().hex[:6]`?

```python
from uuid import uuid4
from django.utils.text import slugify

# [...]
self.code = '{code}-{id}'.format(
    code=slugify(self.code).upper(),
    id=uuid4().hex[:6]
)
```

https://docs.djangoproject.com/en/2.0/ref/utils/#django.utils.text.slugify

File: apps/organizations/models.py [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+        return super().save(*args, **kwargs)
+
+    def get_page(self):
+        """Return the draft CMS page"""
+        try:
+            return self.organization_pages.\
+                select_related('extended_object').\
+                get(extended_object__publisher_is_draft=True).\
+                extended_object
+        except OrganizationPage.DoesNotExist:
```
Comment (jmaupetit) [inline apps/organizations/models.py]: Shouldn't we log something here?

File: apps/organizations/templates/organizations/cms/organization.html [hunk 20]
```diff
@@ -0,0 +1,11 @@
...
+{% block content %}
+<h1>{{ current_page.get_title }}</h1>
+{% with current_page.organizationpage.organization as organization %}
+    {% if organization.logo %}
+    <img src="{{ organization.logo.url }}">
+    {% endif %}
+{% endwith %}
+
{% placeholder "maincontent" %}
+{% endblock content %}
```
Comment (jmaupetit) [inline apps/organizations/templates/organizations/cms/organization.html]: We should discuss templates indentation.

File: apps/organizations/templates/organizations/cms/organization.html [hunk 21]
```diff
@@ -0,0 +1,11 @@
+{% extends "richie/fullwidth.html" %}
+{% load cms_tags %}
+{% block content %}
+<h1>{{ current_page.get_title }}</h1>
+{% with current_page.organizationpage.organization as organization %}
+    {% if organization.logo %}
+    <img src="{{ organization.logo.url }}">
```
Comment (jmaupetit) [inline apps/organizations/templates/organizations/cms/organization.html]: Don't forget the `alt` attribute for a minimal accessibility.

File: apps/organizations/tests/test_admin.py [hunk 22]
```diff
@@ -0,0 +1,102 @@
...
+            response,
+            '<td class="field-view_in_cms"><a href="/en/my-title/" target="_blank">View</a></td>',
+            status_code=200,
+            html=True,
+        )
+
+    def test_organization_list_view_no_page(self):
+        """
+        When the organization is not related to a CMS page, the amdin list view should display a
+        hyphen instead of the link to the associated page
```
Comment (jmaupetit) [inline apps/organizations/tests/test_admin.py]: hyphen or "no-page flag"?

File: apps/organizations/tests/test_admin.py [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+        user = UserFactory(is_staff=True, is_superuser=True)
+        self.client.login(username=user.username, password='password')
+
+        # Get the admin change view
+        url = reverse('admin:organizations_organization_add')
+        response = self.client.get(url, follow=True)
+
+        # Check that the page includes all our fields
+        for field in ['name', 'code', 'logo']:
+            self.assertContains(response, 'id_{:s}'.format(field))
```
Comment (jmaupetit) [inline apps/organizations/tests/test_admin.py]: :medal_sports: :wink:

File: apps/organizations/tests/test_models_organization.py [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+        OrganizationFactory(name='a'*255)
+        with self.assertRaises(DataError):
+            OrganizationFactory(name='a'*256)
+
+    def test_organization_fields_code_normalization(self):
+        """
+        The `code` field should be normalized to improve its uniqueness
+        """
+        organization = OrganizationFactory(code=' r5G ypm 44 ')
+        self.assertEqual(organization.code, 'R5G-YPM-44')
```
Comment (jmaupetit) [inline apps/organizations/tests/test_models_organization.py]: Can we also test with a code containing unicode characters (Emoji, Chinese, ...)?

File: apps/organizations/tests/test_models_organization.py [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+        self.assertEqual(retrieved_page, page)
+        self.assertTrue(retrieved_page.publisher_is_draft)
+
+    def test_organization_get_page_none(self):
+        """
+        Trying to retrieve the page from an organization should return "None" if it
+        does not exist.
+        """
+        organization = OrganizationFactory(name='Sorbonne', code='SOR')
+        self.assertIsNone(organization.get_page())
```
Comment (jmaupetit) [inline apps/organizations/tests/test_models_organization.py]: An empty row is missing at the end of this file.

File: apps/organizations/tests/test_wizards.py [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+        # A parent page to list organizations should pre-exist
+        create_page('Organizations', 'richie/fullwidth.html', 'en', reverse_id='organizations')
+
+        # Submit a title at max length
+        data = {'title': 't' * 255}
+        form = OrganizationWizardForm(data=data)
+        self.assertTrue(form.is_valid())
+        page = form.save()
+        # Check that the slug has been truncated
+        self.assertEqual(page.get_slug(), 't'*200)
```
Comment (jmaupetit) [inline apps/organizations/tests/test_wizards.py]: `self.assertEqual(page.get_slug(), 't' * 200)`?

File: apps/organizations/tests/test_wizards.py [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+        We should not be able to create a CMS Organization Page if the
+        parent page to list organizations was not created
+        """
+        OrganizationFactory()
+        form = OrganizationWizardForm(data={'title': 'My title'})
+        self.assertFalse(form.is_valid())
+        self.assertEqual(form.errors, {'title': [
+            'You must first create an `organization list` page and set its `reverse_id` to '
+            '`organizations`.'
+        ]})
```
Comment (jmaupetit) [inline apps/organizations/tests/test_wizards.py]: Empty row is missing.

File: apps/organizations/admin.py [hunk 28]
```diff
@@ -0,0 +1,35 @@
...
+    # pylint: disable=no-self-use
+    def view_in_cms(self, organization):
+        """
+        Add a link in the admin list view to go to the page of the organization in the CMS
+        """
+        # If the organization is linked to a page on the CMS, we want to display a link that
+        # opens a new tab and displays this page.
+        page = organization.get_page()
+        if page:
+            return '<a href="{url:s}" target="_blank">{anchor!s}</a>'.format(
```
[the remark this answers is not in the document]
Comment (rmoch) [reply apps/organizations/admin.py]: It's a ui decision, I would totally assume
Users of Django admin are us, the dev team
When using keyboard I prefer to close a tab, than hit backspace, because some widget could have take focus, and my right hand is still on the mouse
you ?

[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/organizations/admin.py]: I usually CTR+Click to open the link in a new tab and not having to deal with my current tab history. But it's my choice, my way of organizing the way I browse/work. Nobody forced me to do it that way.

[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/admin.py]: I have no strong opinion here. 
I always use "target=_blank" for links inside django admin, because it is a tool that people use all day long and they most probably don't want you to mess with this tab. 
My feeling is that most websites that are tools work like this (gmail for example)

File: apps/organizations/cms_wizards.py [hunk 29]
```diff
@@ -0,0 +1,99 @@
...
+            cleaned_data['slug'] = slugify(cleaned_data['title'])[:200]
+
+    def clean_title(self):
+        """
+        Check that the organizations list page exists. Do it in a `clean_title` and not as a
+        general error because the wizard page does not display them...
+        """
+        if not Page.objects.filter(
+                reverse_id=ORGANIZATIONS_PAGE_REVERSE_ID,
+                publisher_is_draft=True,
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/cms_wizards.py]: pylint was complaining. It can not be on the same level as below's line "raise..."

File: apps/organizations/models.py (hunk 17 above)
Comment (sampaccoud) [reply apps/organizations/models.py]: I'm pro uuidfield too.

File: apps/organizations/models.py [hunk 30]
```diff
@@ -1,55 +1,91 @@
...
+    page to present the organization.
+    """
+    name = models.CharField(_("name"), max_length=255)
+    code = models.CharField(
+        _("code"),
+        db_index=True,
+        max_length=100,
+        unique=True,
+        null=True,
+        blank=True,
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/models.py]: The idea is that we (FUN) should set codes, not the organization people. So the wizard could be filled by the organization themselves and we later come and add a code.

File: apps/organizations/models.py (hunk 18 above)
Comment (sampaccoud) [reply apps/organizations/models.py]: what's the idea of uuid ? We want to be able to set the code because there are already existing codes. The idea here was just to normalize them a bit.

File: apps/organizations/models.py [hunk 31]
```diff
@@ -1,55 +1,91 @@
...
+        """
+        Try to normalize the code a little for better uniqueness:
+            - all upper case,
+            - strip leading and trailing spaces,
+            - replace other spaces by a hyphen.
+        TODO: improve uniquess (find related Django blog post)
+        """
+        if self.code:
+            self.code = self.code.upper().strip().replace(' ', '-')
+        return super().save(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/models.py]: but we will have slugs... so can we normalize them a bit ?

File: apps/organizations/tests/test_models_organization.py [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+        The `name` field should be required
+        """
+        with self.assertRaises(IntegrityError):
+            OrganizationFactory(name=None)
+
+    def test_organization_fields_name_max_length(self):
+        """
+        The `name` field should be limited to 255 characters
+        """
+        OrganizationFactory(name='a'*255)
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/tests/test_models_organization.py]: I think testing that it works with the max 255 and not with 256.

File: apps/organizations/admin.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/organizations/admin.py]: Let's ask our users then!

File: apps/organizations/models.py (hunk 30 above)
[the remark this answers is not in the document]
Comment (rmoch) [reply apps/organizations/models.py]: 🤔

File: apps/organizations/tests/test_wizards.py (hunk 26 above)
Comment (sampaccoud) [reply apps/organizations/tests/test_wizards.py]: done

File: apps/organizations/tests/test_wizards.py (hunk 27 above)
Comment (sampaccoud) [reply apps/organizations/tests/test_wizards.py]: weird because there is an empty row.

Comment (jmaupetit) [reply apps/organizations/tests/test_wizards.py]: Okay my bad. Forget this comment.

File: apps/organizations/models.py (hunk 18 above)
Comment (sampaccoud) [reply apps/organizations/models.py]: yeah using slugify is a great idea.
I had this blogpost in mind when I added the TODO: https://www.b-list.org/weblog/2018/feb/11/usernames/
I was in the plane so did not remember exactly what to do... should I do it now or is it overkill WDYT?

Comment (jmaupetit) [reply apps/organizations/models.py]: Oh yes, I forgot this post! Yes, I think it's a bit overkill as we are not dealing with (potentially malicious) usernames, right?

File: apps/organizations/tests/test_models_organization.py (hunk 24 above)
Comment (sampaccoud) [reply apps/organizations/tests/test_models_organization.py]: Great idea: " r5G yç👷学pm 44 " -> "R5G-YÇ👷学PM-44"

File: apps/organizations/tests/test_admin.py [hunk 33]
```diff
@@ -0,0 +1,102 @@
...
+        url = reverse('admin:organizations_organization_changelist')
+        response = self.client.get(url, follow=True)
+
+        # Check that the page includes all our fields
+        self.assertContains(response, organization.name, status_code=200)
+        self.assertContains(response, organization.code)
+        self.assertContains(
+            response,
+            '<td class="field-view_in_cms"><a href="/en/my-title/" target="_blank">View</a></td>',
+            status_code=200,
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/tests/test_admin.py]: Ok let me remove it.

File: apps/organizations/tests/test_admin.py [hunk 34]
```diff
@@ -0,0 +1,102 @@
...
+            html=True,
+        )
+
+    def test_organization_list_view_no_page(self):
+        """
+        When the organization is not related to a CMS page, the amdin list view should display a
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/tests/test_admin.py]: fixed it.

File: apps/organizations/tests/test_admin.py (hunk 22 above)
Comment (sampaccoud) [reply apps/organizations/tests/test_admin.py]: well spotted!

File: apps/organizations/tests/test_admin.py [hunk 35]
```diff
@@ -0,0 +1,102 @@
...
+
+        # Create an organization linked to a page
+        page = create_page('My title', 'richie/fullwidth.html', 'en')
+        organization = OrganizationFactory()
+        OrganizationPage.objects.create(organization=organization, extended_object=page)
+
+        # Get the admin change view
+        url = reverse('admin:organizations_organization_change', args=[organization.id])
+        response = self.client.get(url, follow=True)
+
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/tests/test_admin.py]: Must be a copy/paste fail because actually it's not necessary. Removed it.

File: apps/organizations/admin.py (hunk 28 above)
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/organizations/admin.py]: And make this issue non-blocking by declaring an issue to push further the discussion?

File: apps/organizations/models.py (hunk 19 above)
Comment (sampaccoud) [reply apps/organizations/models.py]: No it's not abnormal if there is no page. I use the try/except pattern because it is better to ask for forgiveness than permission.

File: apps/organizations/models.py [hunk 36]
```diff
@@ -1,55 +1,91 @@
...
 
-    def __repr__(self):
+    def __str__(self):
         """Human representation of an organization page"""
-        return "Organization page: {key}".format(key=self.organization_key)
+        return '{model}: {name} ({code})'.format(
+            code=self.organization.code,
+            name=self.organization.name,
+            model=self._meta.verbose_name.title(),
+        )
```
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/models.py]: Don't know where I read it but verbose_names should be lowercase as Django takes care of capitalizing them whenever it is required.

File: apps/organizations/templates/organizations/cms/organization.html (hunk 20 above)
Comment (sampaccoud) [reply apps/organizations/templates/organizations/cms/organization.html]: I modified it.

File: apps/organizations/tests/test_models_organization.py (hunk 32 above)
[the remark this answers is not in the document]
Comment (rmoch) [reply apps/organizations/tests/test_models_organization.py]: 👌

File: apps/organizations/models.py (hunk 36 above)
[the remark this answers is not in the document]
Comment (rmoch) [reply apps/organizations/models.py]: 👌

File: apps/organizations/tests/test_models_organization.py [hunk 37]
```diff
@@ -0,0 +1,112 @@
...
+        with self.assertRaises(DataError):
+            OrganizationFactory(name='a'*256)
+
+    def test_organization_fields_code_normalization(self):
+        """
+        The `code` field should be normalized to improve its uniqueness
+        Test with a chinese character, an emoji and a french ç...
+        """
+        organization = OrganizationFactory(code=' r5G yç👷学pm 44 ')
+        self.assertEqual(organization.code, 'R5G-YPM-44')
```
[the remark this answers is not in the document]
Comment (jmaupetit) [reply apps/organizations/tests/test_models_organization.py]: I think we must keep them. You can allow unicode with slugify: `slugify(value, allow_unicode=True)`

File: apps/organizations/templates/organizations/cms/organization.html [hunk 38]
```diff
@@ -2,14 +2,13 @@
 {% load cms_tags %}
 
 {% block content %}
+  <h1>{{ current_page.get_title }}</h1>
 
-<h1>{{ current_page.get_title }}</h1>
+  {% with current_page.organizationpage.organization as organization %}
+    {% if organization.logo %}
+      <img src="{{ organization.logo.url }}">
```
Comment (jmaupetit) [inline apps/organizations/templates/organizations/cms/organization.html]: If you don't mind, can you add an `alt` attribute here?

File: apps/organizations/tests/test_models_organization.py (hunk 37 above)
[the remark this answers is not in the document]
Comment (sampaccoud) [reply apps/organizations/tests/test_models_organization.py]: I don't think we should because then come all the problems of unicode for unity cf ubernostrum blogpost. What we want for these codes is pure ascii.

File: apps/organizations/templates/organizations/cms/organization.html (hunk 38 above)
Comment (sampaccoud) [reply apps/organizations/templates/organizations/cms/organization.html]: Done
